ai_server/server.py

```diff
@@ -1,12 +1,250 @@
+        return response if response else "No response generated."
+        
+    except subprocess.TimeoutExpired:
+        raise Exception(f"Llama.cpp request timed out for model {model}")
+    except subprocess.CalledProcessError as e:
+        stderr_text = ""
+        if e.stderr:
+            try:
+                stderr_text = e.stderr.decode('utf-8')
+            except UnicodeDecodeError:
+                stderr_text = e.stderr.decode('utf-8', errors='replace')
+        raise Exception(f"Llama.cpp failed for {model}: {stderr_text.strip() if stderr_text else 'Unknown error'}")
+    except FileNotFoundError:
+        raise Exception("Llama.cpp CLI not found")
+
+def chat_with_model(model: str, content: str, llama_mode: str = "cli") -> str:
+    """Route chat request based on llama_mode: server (external), cli, or default fallback."""
+    if llama_mode == "server":
+        # Use external llama-server with fallbacks
+        if not LLAMA_SERVER_URL:
+            raise Exception("LLAMA_SERVER_URL environment variable not set for server mode")
+        
+        if is_llamacpp_available(model):
+            try:
+                return chat_with_llama_server_http(model, content)
+            except Exception:
+                # Fallback to llama-cli if server fails
+                try:
+                    return chat_with_llamacpp(model, content)
+                except Exception:
+                    # Final fallback to ollama
+                    return chat_with_ollama(model, content)
+        else:
+            # Model not available in llama.cpp, use ollama
+            return chat_with_ollama(model, content)
+    
+    elif llama_mode == "cli":
+        if is_llamacpp_available(model):
+            try:
+                return chat_with_llamacpp(model, content)
+            except Exception:
+                # Fallback to ollama if llama-cli fails
+                try:
+                    return chat_with_ollama(model, content)
+                except Exception as e_ollama:
+                    available_llamacpp = get_available_llamacpp_models()
+                    raise Exception(f"Model '{model}' failed in both llama-cli and ollama. "
+                                   f"Available llama.cpp models: {available_llamacpp}. "
+                                   f"Ollama error: {str(e_ollama)}")
+        else:
+            # Model not available in llama.cpp, try ollama
+            try:
+                return chat_with_ollama(model, content)
+            except Exception as e:
+                available_llamacpp = get_available_llamacpp_models()
+                raise Exception(f"Model '{model}' not available in llama.cpp and failed in ollama. "
+                               f"Available llama.cpp models: {available_llamacpp}. "
+                               f"Ollama error: {str(e)}")
+    
+    else:
+        # Default
+        return chat_with_model(model, content, "cli")
```

Comment:
Raise an exception here: the user should see that they entered an invalid value for `llama_mode`.